### DISCUSSION TITLE
[Proposal] Shorthand for minimizing IDisposable resource lifetime

### DISCUSSION THREAD

**Discussion by rwdd** [dotnet/csharplang#4738] [Language Ideas]
Coming from the old style using directive where we wrote something like.
```
TypeA varA;
using (Resource resource = this.CreateResource()) 
{
    varA = resource.GetVarA();
}
//...
this.DoSomething(varA);
```
We now can write it like

```
TypeA varA;
{
    using Resource resource = this.CreateResource();
    varA = resource.GetVarA();
}
//...
this.DoSomething(varA);
```

But wouldn't it be better to be able to write it like
```
var varA = (using this.CreateResource()).GetVarA();
//...
this.DoSomething(varA);
```
Or maybe
```
var varA = (using this.CreateResource(), Resource.GetVarA);
//...
this.DoSomething(varA);
```

**Comment by jnm2** [comment]
@rwdd At what point would the instance be disposed in your examples?

**Reply by rwdd** [reply]
Ah sorry, forgot that. As soon as possible. That would be right after `Resource.GetVarA` has returned.

**Reply by CyrusNajmabadi** [reply]
These would then be different semantics as varA would now live longer than the resource.

**Reply by cheong00** [reply]
> These would then be different semantics as varA would now live longer than the resource.

In the first example given, varA will also live longer than the resource.

So as long as varA don't hold anything that should live with **resource**, I think it's okay to have some syntax for this purpose. (Say, one-off RestfulAPI client that can be safely disposed after JSON is retrieved)

**Reply by rwdd** [reply]
that is exactly what i had in mind. Just think of varA as a primitive type. 
The aim is not to write this ugly code with curly brackets and not to have the lifetime of the resource unnecessarily long.
